File: leetcode/problems/problem-specific/110.balanced-binary-tree/110.balanced-binary-tree.go
package main

import "math"

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isBalanced(root *TreeNode) bool {
	_, isBalanced := isBalancedHelper(root)
	return isBalanced
}

func isBalancedHelper(root *TreeNode) (int, bool) {
	if root == nil {
		return 0, true
	}

	leftSubtreeHeight, isLeftSubtreeBalanced := isBalancedHelper(root.Left)
	rightSubtreeHeight, isRightSubtreeBalanced := isBalancedHelper(root.Right)
	return findMax(leftSubtreeHeight, rightSubtreeHeight) + 1, isLeftSubtreeBalanced && isRightSubtreeBalanced && int(math.Abs(float64(leftSubtreeHeight)-float64(rightSubtreeHeight))) <= 1
}

func findMax(a, b int) int {
	if a >= b {
		return a
	}

	return b
}

File: leetcode/problems/sessions/2023-05-01_2023-09-16/110.balanced-binary-tree.py
#
# @lc app=leetcode id=110 lang=python3
#
# [110] Balanced Binary Tree
#

# @lc code=start

# * Recursive DFS Solution | O(n) Time | O(h) Space
# * n -> The number of nodes in the tree | h -> The height of the tree


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def dfs(root):
            if not root:
                return True, 0

            is_left_subtree_balanced, left_node_depth = dfs(root.left)
            is_right_subtree_balanced, right_node_depth = dfs(root.right)
            return (
                is_left_subtree_balanced
                and is_right_subtree_balanced
                and abs(left_node_depth - right_node_depth) <= 1,
                1 + max(left_node_depth, right_node_depth),
            )

        is_tree_balanced, _ = dfs(root)
        return is_tree_balanced


# @lc code=end
